src/payjoin_sender.rs

```
@@ -0,0 +1,266 @@
+					sleep().await;
+					continue;
+				},
+			};
+
+			if response.status() == StatusCode::OK {
+				let response = match response.bytes().await {
+					Ok(response) => response.to_vec(),
+					Err(e) => {
+						log_info!(
+							self.logger,
+							"Payjoin Sender: Error reading polling response: {}",
+							e
+						);
+						sleep().await;
+						continue;
+					},
+				};
+				if response.is_empty() {
+					log_info!(self.logger, "Payjoin Sender: Got empty response while polling");
+					sleep().await;
+					continue;
+				}
+				return Some(response);
+			} else {
+				log_info!(
+					self.logger,
+					"Payjoin Sender: Error sending request, got status code
+					{}",
+					response.status()
+				);
+				sleep().await;
+				continue;
+			}
+		}
+	}
+
+	// This method will finalise the transaction and broadcast it to the network.
+	fn finalise_payjoin_tx(
+		&self, mut payjoin_proposal_psbt: Psbt, original_psbt: &mut Psbt,
+	) -> Result<Txid, Error> {
+		// We need to reintroduce utxo from original psbt, otherwise BDK will not sign and finalise
+		// the transaction.
+		fn input_pairs(
+			psbt: &mut PartiallySignedTransaction,
+		) -> Box<dyn Iterator<Item = (&bdk::bitcoin::TxIn, &mut Input)> + '_> {
+			Box::new(psbt.unsigned_tx.input.iter().zip(&mut psbt.inputs))
+		}
+
+		// get original inputs from original psbt clone
+		let mut original_inputs = input_pairs(original_psbt).peekable();
+		for (proposed_txin, proposed_psbtin) in input_pairs(&mut payjoin_proposal_psbt) {
+			if let Some((original_txin, original_psbtin)) = original_inputs.peek() {
+				if proposed_txin.previous_output == original_txin.previous_output {
+					proposed_psbtin.witness_utxo = original_psbtin.witness_utxo.clone();
+					proposed_psbtin.non_witness_utxo = original_psbtin.non_witness_utxo.clone();
+					original_inputs.next();
+				}
+			}
+		}
+
+		let is_signed = self.wallet.sign_transaction(&mut payjoin_proposal_psbt)?;
```

we should probably return an error here?